### DISCUSSION TITLE
Super Resolution (Supersampling)

### DISCUSSION THREAD

**Discussion by 733737** [OpenRCT2/OpenRCT2#16884] [Feature requests]
Super Resolution is used in thousands of games to enhance detail. Here is a mock up of what OpenRCT2 would look like if you added super resolution: 
before
![original](https://user-images.githubusercontent.com/94884086/160217454-c91c72c0-3690-4e5f-9cae-6512b1bb48e9.png)
after
![mockup](https://user-images.githubusercontent.com/94884086/160217726-168dec5b-0584-4512-8bb1-c58b86d5ac29.jpg)

**Comment by 733737** [comment]
it eliminates pixelation, aliasing, and jagged edges. I apologize for the oversaturation, that was an error on my part... It is very noticeable on the path railing, monorail track, motion simulator, and tree outlines

**Comment by ocalhoun6** [comment]
To me, the 2nd version actually looks much worse. Besides the color issues, it just looks blurred, and it loses a lot of the crisp details.

For example, the roof texture of the entrance/exit buildings almost entirely disappears in the modified version. The trees also lose a lot of detail.

**Comment by 733737** [comment]
it is a mock up, but it's true that it may appear "blurry" because of how the human eyes process information. Edge Enhancement (or sharpening) looks better but destroys detail. Super resolution if implemented properly, does not destroy detail. It would be optional to implement of course, and it would have no effect on color and would not cause ghosting in motion because it is done in real time

**Comment by 733737** [comment]
here it is with no ai and no added color 
![really](https://user-images.githubusercontent.com/94884086/160223780-860e078f-4210-45c8-8e11-fc096d713529.png)
it does not look very good because I used a basic nonalgorithmic super sampler https://en.wikipedia.org/wiki/Supersampling

**Comment by Broxzier** [comment]
Super-sampling, or rather anti-aliasing, doesn't make much sense for non-scaled sprite based applications. When increasing the game's window scale to make it appear zoomed in, another form of scaling could be used. @janisozaur made a [test PR](https://github.com/OpenRCT2/OpenRCT2/pull/2328) to try this out a long time ago.

![](https://camo.githubusercontent.com/f0932ec07a970d0b091b57bb920fa239fd468769c860f1987486906b34f49793/687474703a2f2f692e696d6775722e636f6d2f6f64546f664a382e706e67)

**Reply by IntelOrca** [reply]
While this is true, if you compare a giant screenshot with a normal screenshot that is zoomed out, the giant one is definitely better quality.

**Reply by 733737** [reply]
the giant screenshot renders at r higher resolution than the game renders in real time

**Comment by 733737** [comment]
super resolution imo should be a 4x linear integer scaler, that downsamples to native by 2x, so if it renders pixels isometrically, you will double the native resolution, but it should be 4x since there would be a 4:1 then it would be rendered at 1/2 of upscaled resolution

**Comment by oli414** [comment]
I think I see what you're saying, but I'm not sure this is relevant since RCT2 scales very well with screen resolutions, unlike retro games with a fixed resolution far lower than that of the target display (e.g 600x400).
On a 4k display, it might be nice, though the 2x or 3x scaling using nearest neighbor works quite well for most people.

With that said, I think what you mean is:

Render the game at a small resolution based on some scaling factor, e.g 480p (16:9)

![image](https://user-images.githubusercontent.com/2348094/160300920-b74d33e2-cd5c-494e-9e48-067deb0aa1ec.png)

Scale it up using nearest neighbor (e.g 800%)
Then scale that back down to our display resolution (1080p) but with bilinear sampling

(click on the image to see it in full resolution)
![image](https://user-images.githubusercontent.com/2348094/160300963-2aa55dd0-9b9d-454e-8a87-4f7e6b3d8ffe.png)

That's 2.25x scaling without it looking completely awful

That essentially maintains the pixelated details without the requirement for a power of 2 scaling value. With just nearest neighbour you'd get the odd row of 3 wide pixels instead of 2 wide, this method transitions between pixels more smoothly without appearing blurry.

Differences close-up, this method:
![image](https://user-images.githubusercontent.com/2348094/160301176-2eaf8b55-d0f3-419f-88fb-31bca0a65baf.png)
Nearest neighbour:
![image](https://user-images.githubusercontent.com/2348094/160301181-7ed2ed4b-7657-438b-ac1d-9dd7347f763b.png)

**Comment by 733737** [comment]
I play at 5120x2160 with auto hdr and the internal res says it's around 1600x686 or so which is around 1/3, it would look better imo if this was super sampled to 6k or higher then downsampled to my native 5120x2160

**Comment by 733737** [comment]
most games offer a selection of 2x 4x or 8x supersampling

**Comment by J-PIE-314** [comment]
This request is similar to #15916.
